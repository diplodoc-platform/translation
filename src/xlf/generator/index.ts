export {
    generateTemplate,
    TemplateParameters,
    validParameters as templateValidParameters,
} from './template';
export {generateTransUnit, TransUnitParameters} from './trans-unit';

export * as default from './index';
