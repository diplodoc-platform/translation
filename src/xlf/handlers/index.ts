import {anchors} from './diplodoc/anchors';

const handlers = {
    ...anchors,
};

export {handlers};
export default {handlers};
