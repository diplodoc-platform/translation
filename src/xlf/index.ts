export * as mdXLFRenderer from './renderer/md-xlf/';
export * as generator from './generator';
export * as parser from './parser';

export * as default from './index';
