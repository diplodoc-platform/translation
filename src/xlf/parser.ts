import {XMLValidator} from 'fast-xml-parser';
import cheerio from 'cheerio';

export type TranslationUnitsByID = Map<string, string>;

export type GetTranslationsParameters = {
    xlf: string;
    startID?: number;
};

function parseTranslations(parameters: GetTranslationsParameters): TranslationUnitsByID {
    const {xlf, startID = 1} = parameters;
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const query = cheerio.load(xlf);

    let translations = query('trans-unit').find('target');
    if (!translations) {
        throw new Error('failed to parse trans-units');
    }

    if (translations.length === 0) {
        translations = query('trans-unit');
    }

    const units = new Map<string, string>();

    let i = startID;
    for (const target of translations) {
        let text = '';
        for (const child of target.children ?? []) {
            if (child.type === 'text') {
                text += child.data;
            }
        }

        units.set(String(i++), text);
    }

    return units;
}

function validParameters(parameters: GetTranslationsParameters) {
    const {xlf} = parameters;

    const conditions = [Boolean(xlf), validXML(xlf)];

    return conditions.reduce((a, v) => a && v, true);
}

function validXML(xml: string) {
    const validation = XMLValidator.validate(xml);

    return typeof validation === 'boolean' && validation;
}

export {parseTranslations, validParameters};
export default {parseTranslations, validParameters};
