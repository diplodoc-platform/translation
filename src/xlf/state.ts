import {XLFRulesState} from './rules';

export type XLFRendererState = XLFState & XLFRulesState;
export type XLFState = {
    xlf: {
        id: number;
        indentation: number;
    };
};

export type XLFInitStateParameters = {
    indentation: number;
};

function xlfInitState(parameters: XLFInitStateParameters) {
    return {
        xlf: {
            id: 1,
            indentation: parameters.indentation,
        },
    };
}

export {xlfInitState};
export default {xlfInitState};
