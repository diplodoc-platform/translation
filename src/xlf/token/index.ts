export type XLFToken = XLFTextToken | XLFTagToken;

export type XLFTextToken = {
    type: 'text';
    data: string;
};

export type XLFTagToken = {
    type: 'tag';
    data: string;
    ctype?: string;
    equivText?: string;
};

function isXLFTextToken(token: XLFToken): token is XLFTextToken {
    return token?.type === 'text';
}

function isXLFTagToken(token: XLFToken): token is XLFTagToken {
    return token?.type === 'tag';
}

export {isXLFTextToken, isXLFTagToken};
export default {isXLFTextToken, isXLFTagToken};
