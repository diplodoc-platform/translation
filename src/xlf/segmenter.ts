import {sentenize} from '@diplodoc/sentenizer';

import {transUnit} from 'src/xlf/generator';
import {XLFRendererState} from './state';

function segmenter(content: string, state: XLFRendererState) {
    let rendered = '';

    for (const segment of sentenize(content)) {
        rendered += transUnit.generate({
            source: segment,
            id: state.xlf.id,
            indentation: state.xlf.indentation,
        });

        rendered += '\n';

        state.xlf.id++;
    }

    return rendered;
}

export {segmenter};
export default {segmenter};
