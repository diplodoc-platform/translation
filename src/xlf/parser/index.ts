import assert from 'assert';
import {XMLValidator} from 'fast-xml-parser';
import cheerio, {Cheerio} from 'cheerio';
import {ChildNode, Element, isTag, isText} from 'domhandler';

import {XLFTagToken, XLFTextToken, XLFToken} from 'src/xlf/token';

export type GetTranslationsParameters = {
    xlf: string;
    useSource?: boolean;
};

const selfClosingTags = new Set(['x']);

function parseTranslations(parameters: GetTranslationsParameters): Array<Array<XLFToken>> {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const {useSource = false} = parameters;
    let xlf = parameters.xlf;
    let {targets, success} = selectTargets(xlf);
    if (!success) {
        if (useSource) {
            xlf = replaceSourceWithTarget(xlf);
            ({targets, success} = selectTargets(xlf));
        } else {
            throw new Error('did not find any translations');
        }
    }

    return parseTargets(targets);
}

function replaceSourceWithTarget(xlf: string) {
    return xlf.replace(/<source>/gmu, '<target>').replace(/<\/source>/gmu, '</target>');
}

function selectTargets(xlf: string) {
    let success = true;
    const query = cheerio.load(xlf);

    const targets = query('trans-unit').find('target');
    if (targets.length === 0) {
        success = false;
    }

    return {
        targets,
        success,
    };
}

function parseTargets(targets: Cheerio<Element>) {
    const parsed = new Array<Array<XLFToken>>();
    const ref = {nodes: []};

    for (const target of targets.get()) {
        inorderNodes(target, ref);

        const tokens = nodesIntoXLFTokens(ref.nodes);
        parsed.push(tokens);

        ref.nodes = [];
    }

    return parsed;
}

function inorderNodes(node: ChildNode, ref: {nodes: ChildNode[]}) {
    if (!node) {
        return;
    }

    if (isText(node)) {
        ref.nodes.push(node);
    }

    if (isTag(node)) {
        if (selfClosingTags.has(node.name)) {
            // eslint-disable-next-line no-param-reassign
            node.attribs.nodeType = 'self-closing';
        } else {
            // eslint-disable-next-line no-param-reassign
            node.attribs.nodeType = 'open';
        }
        ref.nodes.push(node);

        let next = node.firstChild;

        while (next) {
            inorderNodes(next, ref);
            next = next.nextSibling;
        }
    }

    if (isTag(node) && !selfClosingTags.has(node.name)) {
        const closeNode = node.cloneNode();
        closeNode.attribs.nodeType = 'close';
        ref.nodes.push(closeNode);
    }
}

function nodesIntoXLFTokens(nodes: ChildNode[]): XLFToken[] {
    const tokens = new Array<XLFToken>();

    for (const node of nodes) {
        if (isTag(node)) {
            const nodeType = node?.attribs?.nodeType;
            assert(nodeType === 'open' || nodeType === 'close' || nodeType === 'self-closing');

            const token: XLFTagToken = {
                type: 'tag',
                data: node.name,
                nodeType,
            };

            const [_, syntax] = node?.attribs?.ctype?.split('-') ?? [];
            if (syntax?.length) {
                token.syntax = syntax;
            }

            const equivText = node?.attribs['equiv-text'];
            if (equivText?.length) {
                token.equivText = equivText;
            }

            tokens.push(token);
        } else if (isText(node)) {
            const token: XLFTextToken = {
                type: 'text',
                data: node.data,
            };
            tokens.push(token);
        }
    }

    return tokens;
}

function validParameters(parameters: GetTranslationsParameters) {
    const {xlf} = parameters;

    const conditions = [Boolean(xlf), validXML(xlf)];

    return conditions.reduce((a, v) => a && v, true);
}

function validXML(xml: string) {
    const validation = XMLValidator.validate(xml);

    return typeof validation === 'boolean' && validation;
}

export {parseTranslations, validParameters};
export default {parseTranslations, validParameters};
