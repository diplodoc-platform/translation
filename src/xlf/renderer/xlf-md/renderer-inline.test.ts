import {XLFMDRenderer} from './index';
import {XLFToken} from 'src/xlf/token';

describe('renders xlf to markdown', () => {
    it('renders plain text', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'text', data: 'Sentence about something. '},
            {type: 'text', data: 'Another sentence.'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });

    it('renders strong wrapped in <g> tags', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'tag', data: 'target', nodeType: 'open'},
            {type: 'text', data: 'Предложение номер '},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'open',
                syntax: 'strong',
                equivText: '**',
            },
            {type: 'text', data: 'один'},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'close',
                syntax: 'strong',
                equivText: '**',
            },
            {type: 'text', data: '.'},
            {type: 'tag', data: 'target', nodeType: 'close'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });

    it('renders em wrapped in <g> tags', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'tag', data: 'target', nodeType: 'open'},
            {type: 'text', data: 'Предложение номер '},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'open',
                syntax: 'em',
                equivText: '*',
            },
            {type: 'text', data: 'один'},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'close',
                syntax: 'em',
                equivText: '*',
            },
            {type: 'text', data: '.'},
            {type: 'tag', data: 'target', nodeType: 'close'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });

    it('renders s wrapped in <g> tags', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'tag', data: 'target', nodeType: 'open'},
            {type: 'text', data: 'Предложение номер '},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'open',
                syntax: 's',
                equivText: '~~',
            },
            {type: 'text', data: 'один'},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'close',
                syntax: 's',
                equivText: '~~',
            },
            {type: 'text', data: '.'},
            {type: 'tag', data: 'target', nodeType: 'close'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });

    it('renders sup wrapped in <g> tags', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'tag', data: 'target', nodeType: 'open'},
            {type: 'text', data: 'Предложение номер '},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'open',
                syntax: 'sup',
                equivText: '^',
            },
            {type: 'text', data: 'один'},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'close',
                syntax: 'sup',
                equivText: '^',
            },
            {type: 'text', data: '.'},
            {type: 'tag', data: 'target', nodeType: 'close'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });

    it('renders samp wrapped in <g> tags', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'tag', data: 'target', nodeType: 'open'},
            {type: 'text', data: 'Предложение номер '},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'open',
                syntax: 'samp',
                equivText: '##',
            },
            {type: 'text', data: 'один'},
            {
                type: 'tag',
                data: 'g',
                nodeType: 'close',
                syntax: 'samp',
                equivText: '##',
            },
            {type: 'text', data: '.'},
            {type: 'tag', data: 'target', nodeType: 'close'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });

    it('renders code wrapped in <x> tag', () => {
        const renderer = new XLFMDRenderer();
        const tokens: Array<XLFToken> = [
            {type: 'tag', data: 'target', nodeType: 'open'},
            {type: 'text', data: 'Предложение номер '},
            {
                type: 'tag',
                data: 'x',
                nodeType: 'open',
                syntax: 'code',
                equivText: '`один`',
            },
            {type: 'text', data: '.'},
            {type: 'tag', data: 'target', nodeType: 'close'},
        ];

        const rendered = renderer.render(tokens);
        expect(rendered).toMatchSnapshot();
    });
});
