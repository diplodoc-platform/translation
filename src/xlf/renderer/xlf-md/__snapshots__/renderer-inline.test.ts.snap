// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders xlf to markdown renders code wrapped in <x> tag 1`] = `"Предложение номер \`один\`."`;

exports[`renders xlf to markdown renders em wrapped in <g> tags 1`] = `"Предложение номер *один*."`;

exports[`renders xlf to markdown renders plain text 1`] = `"Sentence about something. Another sentence."`;

exports[`renders xlf to markdown renders s wrapped in <g> tags 1`] = `"Предложение номер ~~один~~."`;

exports[`renders xlf to markdown renders samp wrapped in <g> tags 1`] = `"Предложение номер ##один##."`;

exports[`renders xlf to markdown renders strong wrapped in <g> tags 1`] = `"Предложение номер **один**."`;

exports[`renders xlf to markdown renders sup wrapped in <g> tags 1`] = `"Предложение номер ^один^."`;
